File: api/src/main/java/eu/suro/locale/Locale.java
package eu.suro.locale;

import lombok.Getter;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;
import org.yaml.snakeyaml.introspector.PropertyUtils;

import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Getter
class Locale {

    private static final Yaml YAML = newSnakeYaml();

    private static Yaml newSnakeYaml() {
        Constructor constructor = new Constructor();
        PropertyUtils propertyUtils = new PropertyUtils();
        propertyUtils.setSkipMissingProperties(true);
        constructor.setPropertyUtils(propertyUtils);

        return new Yaml(constructor);
    }

    private final String name;
    private int size;
    private final Map<String, String> messages = new ConcurrentHashMap<>();
    private final Map<String, List<String>> listMessages = new ConcurrentHashMap<>();

    Locale(String name) {
        this.name = name;
        this.size = 0;
        try {
            loadFromSite("core/");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Загрузка локализации с github
     * @param path - путь до файла
     * @throws IOException - если не удалось загрузить файл
     */
    void loadFromSite(String path) throws IOException {
        URL oracle = new URL("https://raw.githubusercontent.com/hyneo/Localization/main/lang/"+ path + name + ".yml");
        Map<Object, Object> map = YAML.loadAs(new InputStreamReader(oracle.openStream()), LinkedHashMap.class);
        if (map == null)
            return;

        this.size = map.size();
        for (Map.Entry<Object, Object> entry : map.entrySet()) {
            String key = entry.getKey().toString();
            Object data = entry.getValue();

            if (data instanceof List) {
                listMessages.put(key, (List<String>) data);
                continue;
            }

            messages.put(key, data.toString());
        }
    }

    /**
     * Добавление локализации с github
     * @param path - путь до файла
     * @throws IOException - если не удалось загрузить файл
     */
    void addFromSite(String path) throws IOException {
        URL oracle = new URL("https://raw.githubusercontent.com/hyneo/Localization/main/lang/"+ path + name + ".yml");
        Map<Object, Object> map = YAML.loadAs(new InputStreamReader(oracle.openStream()), LinkedHashMap.class);
        if (map == null)
            return;

        this.size = this.size + map.size();
        for (Map.Entry<Object, Object> entry : map.entrySet()) {
            String key = entry.getKey().toString();
            Object data = entry.getValue();

            if (data instanceof List) {
                listMessages.put(key, (List<String>) data);
                continue;
            }

            messages.put(key, data.toString());
        }
    }

}

File: api/src/main/kotlin/eu/suro/locale/Locale.kt
package eu.suro.locale

import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.Constructor
import org.yaml.snakeyaml.introspector.PropertyUtils
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.util.concurrent.ConcurrentHashMap

data class Locale constructor(
        private var name: String
){
    var size:Int = 0
    private val YAML: Yaml = newSnakeYaml()
    var messages: ConcurrentHashMap<String, String> = ConcurrentHashMap()
    var listMessages: ConcurrentHashMap<String, List<String>> = ConcurrentHashMap()
    companion object{
        fun newSnakeYaml(): Yaml{
            val constructor = Constructor()
            val propertyUtils = PropertyUtils()
            propertyUtils.isSkipMissingProperties = true
            constructor.propertyUtils = propertyUtils
            return Yaml(constructor)
        }
    }

    init{
        try {
            loadFromSite("core/")
        }catch (e: IOException){
            e.printStackTrace()
        }
    }

    /**
     * Загрузка локализации с github
     * @param path - путь до файла
     * @throws IOException - если не удалось загрузить файл
     */
    @Throws(IOException::class)
    fun loadFromSite(path: String) {
        val oracle = URL("https://raw.githubusercontent.com/hyneo/Localization/main/lang/$path$name.yml")
        val map = YAML.loadAs(InputStreamReader(oracle.openStream()), LinkedHashMap::class.java) ?: return
        size = map.size
        addMessages(map)
    }

    /**
     * Добавление локализации с github
     * @param path - путь до файла
     * @throws IOException - если не удалось загрузить файл
     */
    @Throws(IOException::class)
    fun addFromSite(path: String) {
        val oracle = URL("https://raw.githubusercontent.com/hyneo/Localization/main/lang/$path$name.yml")
        val map = YAML.loadAs(InputStreamReader(oracle.openStream()), LinkedHashMap::class.java) ?: return
        size += map.size
        addMessages(map)
    }

    private fun addMessages(map: java.util.LinkedHashMap<*, *>){
        for (entry in map){
            val key = entry.key.toString()
            val data = entry.value
            if(data is List<*>){
                listMessages[key] = data as List<String>
                continue
            }
            messages[key] = data.toString()
        }
    }
}